### PR TITLE
Performance improvement - update at most at the "Update rate"

### PR DESCRIPTION
**Pull request by wxmerkt**
This PR achieves two things:

1. It only renders when triggered by the RViz `update` callback: This way, we respect the "Update rate" setting set in the Global Options in RViz. Currently it renders on every incoming message - which for the Anymal controller is 400 Hz - and at ~1.5-2ms per update (due to the pointer resetting and PinocchioLinkUpdater callbacks), this quickly blocks the GPU and results in dropping RViz fps.
2. It catches a few invalid user inputs in the URDF and arrow settings which previously led to RViz crashing

### CODE REVIEW

**Comment by cmastalli** [comment]
Thanks for the improvement :)

File: include/whole_body_state_rviz_plugin/WholeBodyStateDisplay.h [hunk 1]
```diff
@@ -205,7 +212,7 @@ private Q_SLOTS:
   /**@}*/
 
   enum CoMStyle { REAL, PROJECTED }; //!< CoM visualization style
-  bool com_real_;     //!< Label to indicates the type of CoM display (real or
+  bool com_real_; //!< Label to indicates the type of CoM display (real or
```
[the remark this answers is not in the document]
Comment (wxmerkt) [reply include/whole_body_state_rviz_plugin/WholeBodyStateDisplay.h:215]: Thanks! I've just added another commit with the formatting

**Comment by wxmerkt** [comment]
This should be ready to merge as well :-)

File: include/whole_body_state_rviz_plugin/WholeBodyStateDisplay.h [hunk 2]
```diff
@@ -1,14 +1,18 @@
...
+// Copyright (C) 2020-2021, University of Edinburgh, Istituto Italiano di
+// Tecnologia, University of Oxford.
 // Copyright note valid unless otherwise stated in individual files.
 // All rights reserved.
 ///////////////////////////////////////////////////////////////////////////////
 
 #ifndef WHOLE_BODY_STATE_RVIZ_PLUGIN_WHOLE_BODY_STATE_DISPLAY_H
 #define WHOLE_BODY_STATE_RVIZ_PLUGIN_WHOLE_BODY_STATE_DISPLAY_H
 
+#include "whole_body_state_rviz_plugin/ArrowVisual.h"
```
Comment (cmastalli) [inline include/whole_body_state_rviz_plugin/WholeBodyStateDisplay.h:13]: Do we need to move these header before pinocchio stuffs?

I often follow include the internal headers after the global ones.

File: src/WholeBodyStateDisplay.cpp [hunk 3]
```diff
@@ -279,6 +290,7 @@ void WholeBodyStateDisplay::loadRobotModel() {
   if (content.empty()) {
     clearRobotModel();
     setStatus(StatusProperty::Error, "URDF", "URDF is empty");
+    ROS_ERROR("URDF is empty");
```
Comment (cmastalli) [inline src/WholeBodyStateDisplay.cpp:293]: Why do we need to use ROS_ERROR?

Note that we use `setStatus` which should print the error message in Rviz

File: src/WholeBodyStateDisplay.cpp [hunk 4]
```diff
@@ -289,12 +301,22 @@ void WholeBodyStateDisplay::loadRobotModel() {
   if (!descr.initString(robot_model_)) {
     clearRobotModel();
     setStatus(StatusProperty::Error, "URDF", "Failed to parse URDF model");
+    ROS_ERROR("Failed to parse URDF model");
```
Comment (cmastalli) [inline src/WholeBodyStateDisplay.cpp:304]: same question here (see above)

File: src/WholeBodyStateDisplay.cpp [hunk 5]
```diff
@@ -289,12 +301,22 @@ void WholeBodyStateDisplay::loadRobotModel() {
...
-                                     pinocchio::JointModelFreeFlyer(), model_);
+  try {
+    pinocchio::urdf::buildModelFromXML(
+        robot_model_, pinocchio::JointModelFreeFlyer(), model_);
+  } catch (const std::invalid_argument &e) {
+    std::string error_msg = "Failed to instantiate model: ";
+    error_msg += e.what();
+    setStatus(StatusProperty::Error, "Pinocchio-URDFParser",
+              QString::fromStdString(error_msg));
+    ROS_ERROR_STREAM(error_msg);
```
Comment (cmastalli) [inline src/WholeBodyStateDisplay.cpp:317]: and same question here

**Review by cmastalli — COMMENTED** [review]
I have a few questions about this PR.
Additionally, do you know why it is not passing the CI?

File: include/whole_body_state_rviz_plugin/WholeBodyStateDisplay.h (hunk 2 above)
Comment (wxmerkt) [reply include/whole_body_state_rviz_plugin/WholeBodyStateDisplay.h:13]: The formatting did the re-order as far as I can recall

File: src/WholeBodyStateDisplay.cpp (hunk 3 above)
Comment (wxmerkt) [reply src/WholeBodyStateDisplay.cpp:293]: It was just easier to see during debugging than opening the status when it initially segfaulted. Happy to remove

File: src/WholeBodyStateDisplay.cpp (hunk 5 above)
Comment (wxmerkt) [reply src/WholeBodyStateDisplay.cpp:317]: This error was too long to be read properly otherwise I believe

**Comment by wxmerkt** [comment]
I fixed the build issues and changed the CI. It will be live after this PR has been merged. Example output: https://github.com/wxmerkt/state_rviz_plugin/runs/2200889085?check_suite_focus=true

File: src/WholeBodyStateDisplay.cpp (hunk 3 above)
Comment (cmastalli) [reply src/WholeBodyStateDisplay.cpp]: Ok, I see, it is fine to keep it :)

**Review by cmastalli — APPROVED** [review]
Let's merge it if you've finished with it.

**Comment by wxmerkt** [comment]
I'm finished with it :-)

**Comment by cmastalli** [comment]
@wxmerkt -- thanks for improving the plugin.

**Pull request merged**